homeassistant/components/api/__init__.py

```diff
@@ -371,11 +372,16 @@ async def post(self, request, domain, service):
         context = self.context(request)
 
         try:
-            await hass.services.async_call(
-                domain, service, data, blocking=True, context=context
-            )
+            async with timeout(SERVICE_WAIT_TIMEOUT):
+                await shield(
```

Comment:
This means it could run forever and never timeout which might block shutdown

Reply:
If I understand it correctly, it will be cancelled on stage 3 of shutdown. It is only shielded from cancellation from within the `api` component.

Reply:
We would still have to wait the whole timeout of stage 1 and stage 2

Reply:
I think what you want is to shield the current task from cancelation but run the service call in another task

Reply:
suggested change https://github.com/home-assistant/core/commit/56033fa2cbed8e70a3d7cbaa4500a675f8310dbb

Reply:
> I think what you want is to shield the current task from cancelation but run the service call in another task

What I want is for the service call to be independent from the source that calls it.
Please correct me if I am wrong, but it feels redundant, because services are internally run in a separate task anyway, so it's a task in a task.
[hass.services.async_call()](https://github.com/home-assistant/core/blob/1a6184a9aaae04bb28a53ac0fcdf1f75cf6ea9aa/homeassistant/core.py#L1966) calls [ServiceRegistry._execute_service()](https://github.com/home-assistant/core/blob/1a6184a9aaae04bb28a53ac0fcdf1f75cf6ea9aa/homeassistant/core.py#L2087), which creates a new task using [hass.async_add_executor_job()](https://github.com/home-assistant/core/blob/1a6184a9aaae04bb28a53ac0fcdf1f75cf6ea9aa/homeassistant/core.py#L630).

> We would still have to wait the whole timeout of stage 1 and stage 2

This is up to the integration that provides the service. It could be desirable or not for different integrations.
For example, the `automation` integration does not stop any automations while on stage 1 and stage 2, it even allows to start new automations when we enter stage 1 of shutdown, so it is desirable for some automations to run until stage 3. If we use `api` integration to call a service that triggers an automation, it should behave the same way as triggered by other means. We are not affecting the shutdown here, at least not more than a manually executed service.
Other integrations, if desired, will listen to the `homeassistant_stop` event and will cancel their tasks regardless of the source of the service call. The `shield()` will not prevent a task to be canceled from within. Also, `shield()` will not prevent the `api` component itself to shutdown.

Reply:
The task isn't being canceled from within though, it's being canceled by the timeout context manager when it hits the timeout

Reply:
I think it would be fine to use async_create_task instead of making it a background task if you didn't want it to be canceled at shutdown.  

I don't see a good way to avoid creating a task without the shield blocking the inner cancellation when the timeout happens

Reply:
> which creates a new task using hass.async_add_executor_job().

This function doesn't create a task

https://github.com/home-assistant/core/blob/1a6184a9aaae04bb28a53ac0fcdf1f75cf6ea9aa/homeassistant/core.py#L634

The above names the return a task but it's actually a future https://github.com/python/cpython/blob/21c8fbf28dda1f338303295b094b9cc365d101b2/Lib/asyncio/base_events.py#L859

https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor

Reply:
> This function doesn't create a task

Ok, I may be wrong with the terminology, but whatever it is, it is added to `hass._tasks` and is [cancelled](https://github.com/home-assistant/core/blob/1a6184a9aaae04bb28a53ac0fcdf1f75cf6ea9aa/homeassistant/core.py#L865) on stage 3 of shutdown, i.e. at the same time it would have been cancelled if the service was called by the other means. A background task is [cancelled](https://github.com/home-assistant/core/blob/1a6184a9aaae04bb28a53ac0fcdf1f75cf6ea9aa/homeassistant/core.py#L815) a little bit too early.

> The task isn't being canceled from within though, it's being canceled by the timeout context manager when it hits the timeout

Sorry for the misunderstanding. What I mean is that at shutdown it will be cancelled from within and `shield` won't prevent the shutdown. On timeout, the `shield` would be cancelled, but the service would continue as normal.

> I don't see a good way to avoid creating a task without the shield blocking the inner cancellation when the timeout happens

I have tested my solution and it works with both shutdown and shielding the service from either timeout or connection drop.

Reply:
Let's pull in the test change from https://github.com/home-assistant/core/commit/56033fa2cbed8e70a3d7cbaa4500a675f8310dbb#diff-5004e2984d28c766f3ec87a77ce07e69797fea8cc36cd753b5ccb182b14e13a2 as that should better cover what a service call usually does since most of them are not going to be callbacks.

I need to think about it a bit more if this behavior change is ok or if its making an existing risk worse.  I agree that your change will make it match the docs, but I'm not sure its actually safe to do given the service call could run forever instead of being cancelled when the connection is closed.  Maybe thats ok though since we already have this problem with automations.

Reply:
What's I'm having trouble with is that this PR says it restores the behavior of not cancelling the service call when the connection drops, but afaik aiohttp has always cancelled the task when the connection drops (they changed that behavior a few versions ago and we were stuck for a bit before we figured out how to keep the original behavior) so we are talking about a behavior change which is likely to have side effects since service calls didn't need to worry about running forever when called from the api before

Reply:
This is a significant behavior change which is likely to go unnoticed if we don't highlight it in the release notes.

I think I'm ok with merging this as-is with the test updated and if it's labeled a breaking change.

Reply:
Also please add a comment about why it's shielded so someone doesn't opens PR in the future to remove it because they were relying on being able to drop the connection to cancel the service call